feat(swingset): get queues length

mhofman opened the pull request:
closes: #5238
refs: #5176

## Description

This PR tracks the number of messages pending on promise queues as a new kernel stat, and exposes the run-queue + acceptance queue length, as well as the promise queue length in a new controller method, which can be used by cosmic-swingset to decide if messages should be delivered to mailbox or possibly bridge devices.

To make sure the stats are consistent with the rest of the kernel state, this PR also:
- Split the kernel stats into consensus and local sections, saved separately (partially undoes https://github.com/Agoric/agoric-sdk/pull/3625/commits/df8359eca80e28736d294a558ed6c5e3b8b14127)
- Commits / rolls-back the `kernelStats` along kernel state, but avoiding intermediary DB writes

### Security Considerations

The `kernelStats` were previously out of consensus and not thoroughly vetted to be deterministic, or consistent with the kernel state. I believe I have audited all changes to the stats, especially the ones that are non in consensus, and made sure any crank abort or reload would reflect in the stats kept in RAM.

### Documentation Considerations

Description added, stats don't seem to be documented anywhere else than in `metrics.js`.

I have now fully typed the metrics allowing calls like `incStats` to type check that the stat name is valid.

### Testing Considerations

I added unit tests for the stats logic, however nothing tests the content of stats against expectations regarding the rest of the kernel state.

One example are the up/down stats. The base value should always be a pure function of the known state of the kernel, but nothing enforces or verifies that. I've manually reviewed all paths where existing run-queue and acceptance queue, as well as the new promise queue, are modified to ensure stats are tracked appropriately

Putting some of these stats under consensus should quickly highlight non deterministic stats (if any) by the deployment integration test. Swingstore traces in integration should also highlight divergences in non-consensus stats however such consistency is not yet enforced in the integration test due to snapshot hash divergences.

warner commented:
API seems good.. having it depend upon `stats` feels sligtly weird, but maybe that's just because of the not-intended-for-consensus property that you pointed out. OTOH it'd be a shame to duplicate code to get this data.

I'm not 100% sure the stats are managed accurately across crank unwinding and/or replay, I seem to remember them being held in RAM for some period of time, and then getting written to the DB only at the end of a crank, and I remember being nervous about that but decided that since they weren't being used for consensus, it wouldn't matter.

So maybe it *would* be a good idea to track them separately. The run-queue and acceptance-queue are still single keys, we could just JSON.parse and `.length` them when requested (which won't be very often, assuming cosmic-swingset only asks once per block, but OT3H maybe that's a bad assumption to make).

Dunno.

mhofman commented:
I did double check the write logic, and it seemed sane. I will double check the abort crank logic however, as now I'm not sure if we reload the saved object at that point.

That said, I think it should be safe to use for our purpose:
- Accepting messages into the mem-pool is non-consensus in the first place
- Accepting messages into a block is done by the block proposer, which I think means it's also out of consensus in the sense other nodes would accept the proposer's decision?

I'm happy to move a subset of stats explicitly into consensus. Not sure the best approach for that is. I already have reservations about all the transient serializations and deserializations happening during a crank, and moving stats under that mechanism doesn't feel quite right.

mhofman commented:
I did find issues related to crank reverting, and when stats are loaded or first initialized. I have now aligned them with other store entries, and moved them in consensus, including participating in the crankhash. They are however still only written down at the end of the crank to avoid DB churn.

warner commented:
Yeah, I'm dubious about enforcing consensus of kernelStats. It would prevent us from adding new stats in minor kernel updates (those updates would have to be deployed at exactly the same time across all replicas), as well as stats on things that really are replica-local (like the number of times the kernel has been rebooted, or heap snapshot activity, or some sort of cache instrumentation).

Until now (where we want to use some stats for backpressure), these stats have been entirely for telemetry/debugging purposes. Our current plan for backpressure is to use it in a non-consensus fashion (mempool inlet control).

OTOH (and @michaelfig can probably clarify), it might be easy to accidentally (or intentionally) implement the cosmic-swingset -side backpressure queue-depth check in a way that *is* part of consensus (at the beginning of DeliverTx, causing a txn error if the block proposer allows a swingset message in despite the queues being too full).

Part of me wants two separate kernelStats, one local (with most of the stats), and the other in consensus (only containing the queue depths). Then we don't need to worry about cosmic-swingset accidentally relying upon something that swingset doesn't guarantee consensus about. But having two sets of stats feels like bloat, especially if they both need `initializeStats` and updates after each crank.

Maybe there's a deeper problem with thinking of SwingStore as an acceptable place to hold non-consensus data at all, and we should provide the kernel with a secondary non-consensus storage tool instead of using `local.*` to distinguish them. If we ever wanted to try (again) to store swingset data in the cosmos-sdk IAVL tree, the `local.` trick wouldn't work, and we'd definitely need a secondary DB. But that sounds like far more work than we can do for MN-1.

Hrmph. So the short-term options are:
* 1: include kernelStats in consensus
  * risks a MN-1 consensus failure if one of our stats is less stable than we thought (I'm looking suspiciously at `kernelDevices`, probably ok but hmm)
  * adding new non-consensus stats would require more work in the future (but maybe YAGNI)
* 2: split into `kernelStats` and `local.kernelStats`
  * somewhat less risk
  * more work now, kinda annoying bloat

@FUDCo any opinions here?

mhofman commented:
> Part of me wants two separate kernelStats, one local (with most of the stats), and the other in consensus (only containing the queue depths). Then we don't need to worry about cosmic-swingset accidentally relying upon something that swingset doesn't guarantee consensus about. But having two sets of stats feels like bloat, especially if they both need `initializeStats` and updates after each crank.

I think 2 stats is right when we need them, and when I audited the current stats, they all seemed to be based on consensus state already, so I didn't see the harm in moving them all into consensus. I'm not sure 2 different keys, one prefixed by `local.` is that much bloat.

> * I'm looking suspiciously at `kernelDevices`

If we believe this may be different between validators, I'm happy to review, but the devices are already stored under the `device.` key in the store, so they're already part of consensus.

FUDCo commented:
> @FUDCo any opinions here?

I'd also go with option 2.  I think it's proper to distinguish between stats that are kept for performance monitoring, tuning, and debugging vs. stats that are part of the operational logic of the chain.

mhofman commented:
@warner @FUDCo PTAL

mhofman commented on packages/SwingSet/src/kernel/kernel.js:
```diff
@@ -1645,6 +1643,27 @@ export default function buildKernel(
       return kernelKeeper.getActivityhash();
     },
 
+    // While stats are local and technically not part of concensus state, these
```
That comment has been removed already.

mhofman commented on packages/SwingSet/test/test-state.js:
```diff
@@ -701,6 +703,22 @@ test('meters', async t => {
   t.deepEqual(k.getMeter(m3), { remaining: 'unlimited', threshold: 5n });
 });
 
+const makeTestCrankHasher = k => {
```
I'd be happy to revert the string serialization part.

mhofman replied:
Reverted

mhofman commented:
@warner, I addressed your feedback in e16a9a3972f419ad7b9a43a10d278f6b4b60aeb4 then squashed the PR in a way that hid all the back and forth in this PR for a clean merge, while preserving individual commits that make sense: [`e16a9a3..a516a0e`](https://github.com/Agoric/agoric-sdk/compare/e16a9a3972f419ad7b9a43a10d278f6b4b60aeb4..a516a0e0a7c1a7815b855fb75df712b532ac6a4a)
